index.js

```diff
@@ -470,21 +494,45 @@ const createFocusTrap = function (elements, userOptions) {
         return this;
       }
 
-      updateTabbableNodes();
-
       state.active = true;
       state.paused = false;
       state.nodeFocusedBeforeActivation = doc.activeElement;
 
-      const onActivate =
-        activateOptions && activateOptions.onActivate
-          ? activateOptions.onActivate
-          : config.onActivate;
-      if (onActivate) {
-        onActivate();
+      const getOption = (optionName) => {
+        return activateOptions && activateOptions[optionName] !== undefined
+          ? activateOptions[optionName]
+          : config[optionName];
+      };
+
+      const onActivate = getOption('onActivate');
+      const onSuccessfulActivation = getOption('onSuccessfulActivation');
+      const checkCanActivate = getOption('checkCanActivate');
+
+      if (checkCanActivate) {
+        if (onActivate) {
+          onActivate();
+        }
+
+        state.containers.forEach((container) => {
```

Comment:
It's good you're thinking about the containers, but I think this will get really messy because one container may be ready and another not? What if you have 3 and two are delayed for some reason? One is ready, and the other isn't, but it'll be a race condition as it's currently implemented. You could try to manage all this, but it would get complex because of the asynchronicity.

How about we change `checkCanActivate()` to take the list of containers instead of being called once for each container?

Then the complexity of managing 2 independently delayed containers is pushed on to the consumer, who will always have better knowledge of how they want to manage that, and we don't have to deal with it here (nor do we have to test for it).

Reply:
I've tried to implement this but it's going to need a rather large refactor of the whole project for it to work since `updateTabbableNodes()` and `addListeners()` run across all containers all at once. You can't run `updateTabbableNodes()` based on a specific container.

This is an extremely niche use case. For initial release of the feature can we perhaps restrict the activation check to maybe just the first container in the list instead of all the containers?

This should cover 90% or more of the use cases.

You can see my failed attempt in this commit:
https://github.com/focus-trap/focus-trap/pull/407/commits/5a8a7c29eb7cd91e2aafa50cf7b0507f549d1362

Reply:
I think you may have misunderstood what I was asking to change here with the multiple containers: I don't want `delayFocusTrapActivation()` to do anything with the containers. It should treat it as a unit and just pass the whole list to `checkCanActivate()` and let the consumer deal with the list as they see fit. That way, whether they're concerned about one or multiple containers, they can figure out how they can to handle this.

Now, here is what I think the code here should look like (remember from my other comment, `updateTabbableNodes()` is not the method that ultimately sets focus to the configured/defaulted "initial focus node" (see `getInitialFocusNode()`) -- it's `addListeners()` that does that).

Also note that, in order to keep the original code as unmodified as possible, I think `checkCanActivate` should be renamed `checkCanFocus`, because that's really what we're offering to delay. Otherwise, we'd have to also move the point at which `state.active = true; state.paused = false; state.nodeFocusedBeforeActivation = doc.activeElement` are called (moving that last assignment, especially, could have repercussions we don't know about), and I'd prefer to be as surgical as possible here.

```javascript
    activate(activateOptions) {
      if (state.active) {
        return this;
      }

      updateTabbableNodes();

      state.active = true;
      state.paused = false;
      state.nodeFocusedBeforeActivation = doc.activeElement;

      const onActivate = getOption(activateOptions, 'onActivate');
      const onPostActivate = getOption(activateOptions, 'onPostActivate');
      const checkCanFocus = getOption(activateOptions, 'checkCanFocus');

      if (onActivate) {
        onActivate();
      }

      const finishActivation = () => {
        addListeners();
        if (onPostActivate) {
          onPostActivate();
        }
      };

      if (checkCanFocus) {
        // would be even more succinct to use `finally(finishActivation)` but I'm leary
        //  of someone on some old version of some browser that doesn't support
        //  it since it wasn't part of the original Promises spec
        checkCanFocus(state.containers).then(finishActivation, finishActivation);
        return this;
      }

      finishActivation();
      return this;
    },
```

And now we don't even need a `delayFocusTrapActivation()` helper function!

I've also been thinking that `checkCanFocus()` should return a promise (as you can see above) and completely do away with the complexity of any sort of timer. The consumer who handles `checkCanFocus()` is then free to use whatever method they want, at whatever frequency they way, to determine the best time to activate the trap, and then resolve the promise. `focus-trap` will just wait indefinitely until the promise is resolved (or rejected) before it finishes the activation sequence and sets the initial focus.

We would need to document the fact that the `checkCanFocus` is not supported on any version of IE because of the use of a Promise (which is fine since it's new, and all but IE 11 have EOL'ed already, and this is a niche feature anyway; IE 11 EOLs on June 15, 2022; can't wait!), and also that rejecting the promise will not prevent the initial focus from taking place. The promise only serves as a way to delay setting that focus.

And now we end-up with a bonus `onPostActivate()` that someone can use to be notified _after_ the initial focus has been set and all the activation ceremonies have been done.

Reply:
I can try this method. 

I don't think `updateTabbableNodes()` can be called this early, particularly on the animated example, but I can try.

We might have to move it into `finishActivation()` and document it as a possible minor low risk breaking change. 99% of projects won't break.

Reply:
Let's just call `updateTabbableNodes()` a second time in the finish handler -- if focus was delayed:

```javascript
      const finishActivation = (focusDelayed) => {
        if (focusDelayed) {
          updateTabbableNodes();
        }
        addListeners();
        if (onPostActivate) {
          onPostActivate();
        }
      };

      ...

      checkCanFocus(state.containers).then(finishActivation, finishActivation.bind(null, true));
```

This way, we only call it the one time, and at the same place, as before if `checkCanFocus` is not provided (default); otherwise, we call it again "just to be sure", once we're allowed to set focus (and we still call it before setting focus, like we were doing before).

Reply:
I've implemented what you have suggested plus I also added onDeactivate equivalents for when the trigger is the thing that is animated (much less common use case but I can imagine it being a thing).

Since I added onDeactivate equivalents I kept the `checkCanActivate` name for symmetry with `checkCanDeactivate`. 

`checkCanDeactivate` doesn't fully stop the deactivation process though, more like it just holds off on redirecting focus back to the trigger button so `checkCanDeactivate` might not be the best name for it 🤔

Reply:
Maybe `checkCanFocusTrap` and `checkCanFocusTrigger` would be better names....

Yeah I think I'll rename them to those names.

Reply:
Renamed to `checkCanFocusTrap` and `checkCanReturnFocus`